zallet/src/components/json_rpc/methods/list_addresses.rs

```diff
@@ -0,0 +1,222 @@
+    transparent: Option<TransparentAddresses>,
+
+    /// Each element in this list represents a set of diversified addresses derived from a
+    /// single IVK.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    sapling: Vec<SaplingAddresses>,
+
+    /// Each element in this list represents a set of diversified Unified Addresses
+    /// derived from a single UFVK.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    unified: Vec<UnifiedAddresses>,
+}
+
+impl AddressSource {
+    fn empty(source: &'static str) -> Self {
+        Self {
+            source,
+            transparent: None,
+            sapling: vec![],
+            unified: vec![],
+        }
+    }
+
+    fn has_data(&self) -> bool {
+        self.transparent.is_some() || !self.sapling.is_empty() || !self.unified.is_empty()
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct TransparentAddresses {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    addresses: Vec<String>,
+
+    #[serde(rename = "changeAddresses")]
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    change_addresses: Vec<String>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct SaplingAddresses {
+    #[serde(rename = "zip32KeyPath")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    zip32_key_path: Option<String>,
+
+    addresses: Vec<String>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct UnifiedAddresses {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    seedfp: Option<String>,
+
+    /// The ZIP 32 account index.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    account: Option<u32>,
+
+    addresses: Vec<UnifiedAddress>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub(crate) struct UnifiedAddress {
+    /// The diversifier index that the UA was derived at.
+    diversifier_index: u128,
+
+    /// The receiver types that the UA contains (valid values are "p2pkh", "sapling", "orchard").
+    receiver_types: Vec<String>,
+
+    /// The unified address corresponding to the diversifier.
+    address: String,
+}
+
+pub(crate) fn call(wallet: &WalletConnection) -> Response {
+    let mut imported_watchonly = AddressSource::empty("imported_watchonly");
+    let mut mnemonic_seed = AddressSource::empty("mnemonic_seed");
+
+    for account_id in wallet
+        .get_account_ids()
+        .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+    {
+        let account = wallet
+            .get_account(account_id)
+            .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+            // This would be a race condition between this and account deletion.
+            .ok_or(RpcErrorCode::InternalError)?;
+
+        let addresses = wallet
+            .list_addresses(account.id())
+            .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?;
+
+        let mut transparent_addresses = vec![];
+        let mut transparent_change_addresses = vec![];
+        let mut transparent_ephemeral_addresses = vec![];
```

Comment:
We didn't have ephemeral addresses in `zcashd`, so we need to decide where in the output to place them. Note that for historic reasons, the external and change addresses in the `transparent` block of a source are not subgrouped by account. We could bucket all ephemeral addresses together next to them, or we could create a new part of the structure for them.

Reply:
Ephemeral addresses are explicitly a per-account concern, so we should not group them with the historic addresses.

Reply:
> Maybe add a parameter to the RPC to split out {external, change, ephemeral}, but bucket them together by default?

We already put per-account transparent addresses under the correct per-account section. There will never be any ephemeral addresses associated with the legacy account; we don't need to group them together.

> Note that for historic reasons, the external and change addresses in the transparent block of a source are not subgrouped by account.

I thought that that was _only_ true for the legacy random source? If we failed to group per-account transparent addresses with the rest of the account's addresses, that was a major miss.

Reply:
AFAICT from the `zcashd` impl, we did indeed fail to group per-account transparent addresses with the rest of the account's addresses. PR that added this logic is here: https://github.com/zcash/zcash/pull/5681

Reply:
I'm fixing this by adding a new `derivedTransparent` array to each source. `transparent` will only contain transparent addresses for which we do not know derivation information.